### ISSUE TITLE
DB dupe key error in activewatcher table (due to in-memory not having those constraints)

### ISSUE THREAD

**Issue by wdoekes** [OpenSIPS/opensips#1662]
**OpenSIPS version you are running**

```
$ git branch -v | grep '^[*]'
* master       3bce368ce mem: add rpm cli param for selecting allocator
```

Actually, the problem happens in 2.4 for us. But any version I've seen will have the same issue.

**Describe the bug**

In certain situations, writing the activewatcher records to the database creates duplicate key errors; if the client device reuses callid/fromtag.

This causes the DB sync (db_update_period) to stop. In the worst case, this causes a major backlog to the DB updating and SUBSCRIBE expiry handling.

DB sync will stop here (in `update_db_subs`):
```c
if(dbf->insert(db,query_cols,query_vals,n_query_cols )<0)
{
        LM_ERR("unsuccessful sql insert\n");    
        if(!no_lock)
                lock_release(&hash_table[i].lock);
        return ;
```
And that will in turn create trouble with record expiry, which is also handled by `update_db_subs`.

**Describe the bug verbosely**

The database schema says:
```xml
$ sed -e '/<index>/,/<\/index>/!d' db/schema/pr_active_watchers.xml 
    <index>
        <name>active_watchers_idx</name>
        <colref linkend="presentity_uri"/>
        <colref linkend="callid"/>
        <colref linkend="to_tag"/>
        <colref linkend="from_tag"/>
        <unique/>
    </index>
```
But the in-memory representation of the activewatcher does not have this constraint:

- it simply does an `insert_shtable` when a new SUBSCRIBE arrives

This means that one can create multiple activewatcher entries in memory, but when writing the entries to the DB, the writing will halt and abort. When `fallback2db` is not set, this will happen in `update_db_subs`.

**To Reproduce**

1. Making it easier to reproduce, is by reducing the 16 bits of "random" tag to 4 bits. These bits will create the "random" `to_tag`, the only part of the unique constraint created by opensips.
```patch
--- a/crc.c
+++ b/crc.c
@@ -229,6 +229,7 @@ void crcitt_string_array( char *dst, str src[], int size )
                }
        }
        ccitt = ~ccitt;
+       ccitt |= 0xFFF;  // XXX: increase chance of collisions
        if (int2reverse_hex( &dst, &str_len, ccitt )==-1) {
                /* bug ... printed ccitt value longer than CRC32_LEN */
                LM_CRIT("string conversion incomplete\n");
```
(This is of course cheating. But we've had these collisions several times now with the 16 bit tag. It's not a theoretical issue only.)

2. Start making new initial SUBSCRIBEs to the presence module. Make sure you reuse the `URI`, `callid`, `from_tag` (the `to_tag` will be generated by opensips, and incidentally collide with an earlier one). A bit of SIPp XML might come in handy:
```xml
  <send retrans="500" start_txn="subscribe">
    <![CDATA[

      SUBSCRIBE sip:[watchee]@[remote_ip]:[remote_port] SIP/2.0
      Via: SIP/2.0/[transport] [local_ip]:[local_port];branch=[custom_branch2] 
      From: sipp <sip:[service]@[local_ip]:[local_port]>;tag=[ftag]
      To: <sip:[watchee]@[remote_ip]:[remote_port]>
      [authentication]
      Call-ID: [call_id]
      CSeq: [cseq] SUBSCRIBE
      Contact: "sipp" <sip:[service]@[local_ip]:[local_port]>
      Max-Forwards: 70
      Event: dialog
      Expires: 900
      Accept: application/pidf+xml
      Allow-Events: presence
      Content-Length: 0

    ]]>
  </send>
```
3. Subscribe (successfully) a couple of times.

4. Observe how the SQL backend will start producing duplicate key errors:
```
CRITICAL:db_mysql:wrapper_single_mysql_stmt_execute: driver error (1062):
  Duplicate entry '...' for key 'presentity_uri'
```

**Expected behavior**

I'm not sure. Either:

- the table in the DB should lose the UNIQUE constraint and happily accept a second record;

or:

- the in-memory insert should be overwritten with the new record (and existing records should be wiped from the DB / overwritten).

In either case: the DB should behave the same as the in-memory representation.

I did not begin working on a fix yet, mostly because I'm not sure which direction we want to take. I _think_ I would favor the overwrite exising representation approach:

(modules/presence/subscribe.c)
```c
int update_subscription(struct sip_msg* msg, subs_t* subs, int init_req)
{
...
        if(init_req ==0) /*if a SUBSCRIBE within a dialog */
        {
...
        }
        else
        {
...
// here, add code to check whether the record exists, before we blindly insert..
// and then overwrite instead of inserting
                        subs->expires += expires_offset;
                        if(insert_shtable(subs_htable,hash_code,subs)< 0)
                        {
```

What do you say?

**Workarounds**

A workaround could be to set `fallback2db`. That causes new SUBSCRIBEs to get inserted into the activewatchers table immediately (also causing failure for duplicate keys), but then the DB expiry routines will not be affected because the record is now flagged as NO_UPDATEDB_FLAG.

**Comment by bogdan-iancu** [comment]
@wdoekes , thanks for your detailed report.
And I see here two issues actually.

First, the flush to DB  (the inserts) should not stop upon a failure -> the faulty record should be skipped so the rest of the records will have a chance to be flushed to DB.

The second is about the constraint - which I think it is good to have, at DB and code level. You cannot have 2 different subscription within the same SIP dialog (to_tag, from_tag, callid). So, if another SUBSCRIBE is received with the same dialog ID, it should overwrite the existing one.
But as I see it, the problem (correct me if I'm wrong) is how the To-tag is generated by OpenSIPS when handling the SUBSCRIBE. As per your indication, the SUBSCRIBE's you sent are initial ones (without a to-tag). So, it order to get a dialog collision, opensips must generate the same to-tag for both of them - and this is a fault  in the `sl` module (which generates the to_tag).
Is this correct ?

**Comment by bogdan-iancu** [comment]
@wdoekes - any feedback on above ?

**Comment by wdoekes** [comment]
Hi @bogdan-iancu!

Thanks for picking this up :)

> And I see here two issues actually.

Actually, you see three:

> First, the flush to DB (the inserts) should not stop upon a failure

> The second is about the constraint [...] it should overwrite the existing one

> [...] the problem [...] opensips [generates] the same to-tag

Your conclusions are all correct.

**To-tag collisions**

The problems starts with the colliding to-tag. I haven't figured out when this happens, but it was one customer in particular that triggered this more than once.

So while it _could_ just be bad luck (a collision in 16 bits is not _that_ rare), it could also be something that's reproducible seeing that the to-tag is not random, but CRC-based on the contents of the request.

A quick fix could be to use more bits for the CRC. Adding in a random value might not be a good idea, as other opensipsen might depend on tags being generated in a certain way (do they?).

(I consider this a good improvement for opensips overall, but not mandatory to fix this particular issue.)

**The constraint**

That should be relatively simple to fix now that you've decided on overwriting the in-memory version. Care must be taken so the DB-copy is deleted (or overwritten when re-inserting the new one).

(I consider this mandatory.)

**Don't stop processing on single row failure**

Also relatively simple to fix, by not returning immediately after an unsuccessful sql insert.

(I consider this a good safety net, although the problem should not occur if the above is fixed. And in fact, if there is a different cause for the failure, we might not want to continue with *all* inserts anyway..)

**Comment by wdoekes** [comment]
> So while it could just be bad luck (a collision in 16 bits is not that rare), it could also be something that's reproducible seeing that the to-tag is not random, but CRC-based on the contents of the request.

Note that our to-tag was generated behind another opensips (customer -> opensips -> opensips). So the customer was not in full control of the packet contents. The SBC-opensips added its Via first.

In a customer->opensips scenario, creating a collision on purpose would have been trivial, I think.

**Comment by bogdan-iancu** [comment]
For the problem (3), I just committed a fix, see  d4c46def9155c89c3dce3e1b74a93a05448b6fa3 , so we still have the (1) the tag collision and (2) aligning the DB & mem constraints.

For (1) - the TO tag is computed as a CRC over (a) port in VIA, (b) address in VIA and (c) branch param in VIA.
The idea is to ensure, even in stateless mode, the same TO-tag for any reply of the same transaction. 

Now, do you see a potential collision between those 3 elements (in your tests), or an issue in the CRC function ?

**Comment by wdoekes** [comment]
> For the problem (3), I just committed a fix

Nice :+1: 

> Now, do you see a potential collision between those 3 elements (in your tests), or an issue in the CRC function?

- Imagine I put my presence server directly on the internet. Then the customer-device is free to re-use the via-addr + via-port + via-branch. That would allow him to trigger this problem intentionally.

- In our case however, there is (at least) one OpenSIPS machine in front, so the via-branch is not (intentionally) predictable. However, the via-address and via-port are likely always the same.

- So, that brings us to: is the via-branch generation in the front-opensips random enough; or does that create unintentional collisions in the CRC functions in the presence-opensips?

- As for the CRC function itself: I don't have any expertise in that area.

Unfortunately I haven't been able to capture the previous SUBSCRIBEs by the devices that triggered the duplicate key. So I cannot point to the exact cause of the duplicate CRC yet. (I'll go do another dig now, to see if someone stored the evidence after the latest occurrence.)

**Comment by bogdan-iancu** [comment]
@wdoekes, a quick question - your opensips front-end does a stateful or stateless relaying to the presence instance ?
If statefull, the TM will generate the VIA branch  (the one involved in CRC on the presence instance). If stateless, opensips will re-use the received VIA branch.

**Comment by wdoekes** [comment]
Stateful

**Comment by bogdan-iancu** [comment]
OK, in this case the CRC16 gives a really poor dispersion, even if the input is different (different incoming VIA branches). 
As you can reproduce this, if I will send you in the next days a patch to switch from CRC16 to MD5 128 for To-tag calculation, would you give it a try ?

**Comment by wdoekes** [comment]
Certainly :)

**Comment by bogdan-iancu** [comment]
could you try this
[to_tag_md5.patch.txt](https://github.com/OpenSIPS/opensips/files/3228524/to_tag_md5.patch.txt) - it should fit any opensips version

**Comment by wdoekes** [comment]
This might pose a problem:
```
CREATE TABLE active_watchers (
    id INT(10) UNSIGNED AUTO_INCREMENT PRIMARY KEY NOT NULL,
    presentity_uri CHAR(255) NOT NULL,
    watcher_username CHAR(64) NOT NULL,
    watcher_domain CHAR(64) NOT NULL,
    to_user CHAR(64) NOT NULL,
    to_domain CHAR(64) NOT NULL,
    event CHAR(64) DEFAULT 'presence' NOT NULL,
    event_id CHAR(64),
    to_tag CHAR(64) NOT NULL,
```
^-- 32+1+32 being larger than 64

I can fix that locally.. but I may need to browse some to find if there are more of those constraints.

**Comment by bogdan-iancu** [comment]
Yes please do the local change for the moment, just to see if it works

**Comment by bogdan-iancu** [comment]
@wdoekes, here is a new patch that preservers the len of the generated TAGs  [to_tag_md5.patch2.txt](https://github.com/OpenSIPS/opensips/files/3229053/to_tag_md5.patch2.txt). The idea is to use CRC for the fix part (not so important) and MD5 for the variable part. A small trick to keep the len under control, but improve the dispersion of the tag values.

**Comment by wdoekes** [comment]
```
@@ -82,10 +82,11 @@ inline static void init_tags( char *tag, char **suffix,
 	src[2].s=si?si->port_no_str.s:"";
 	src[2].len=si?si->port_no_str.len:0;
 
+	crcitt_string_array( tag, src, 3);
 	MD5StringArray( tag, src, 3 );  // <-- this looks like it should be removed as well
 
-	tag[MD5_LEN]=separator;
-	*suffix=tag+MD5_LEN+1;
+	tag[CRC16_LEN]=separator;
+	*suffix=tag+CRC16_LEN+1;
 }
 ```

**Comment by bogdan-iancu** [comment]
right :D - see  updated [to_tag_md5.patch3.txt](https://github.com/OpenSIPS/opensips/files/3229762/to_tag_md5.patch3.txt)
LEt me know the results of your testing ;)

**Comment by bogdan-iancu** [comment]
@wdoekes , the latest patch was uploaded on master and 3.0. And there is no need for a DB schema change.

**Comment by bogdan-iancu** [comment]
@wdoekes, any chance in testing your case ? I'm really curios to find if it solves the problem (or not)

**Comment by wdoekes** [comment]
> I'm really curious to find if it solves the problem

Well yes. So am I.

(Un)fortunately the bug doesn't trigger every day; it's more like once every couple of weeks.

I'll need to roll a new version and get it through the testing pipeline and into production -- and then wait a bit -- before we can confirm that everything is indeed fixed. But I'm more confident than not that the md5-fix is the right solution.

I'll keep you posted.

**Comment by bogdan-iancu** [comment]
"I'm more confident than not that the md5-fix is the right solution"

Not sure I understand that ^^^ statement - are you or not confident in 
the current fix ?

Regards,

Bogdan-Andrei Iancu

OpenSIPS Founder and Developer
   https://www.opensips-solutions.com
OpenSIPS Summit 2019
   https://www.opensips.org/events/Summit-2019Amsterdam/

**Comment by wdoekes** [comment]
>> "I'm more confident than not that the md5-fix is the right solution"

> Not sure I understand that ^^^ statement - are you or not confident in the current fix ?

Sorry about the confusing language. It literally says:

I'm _more_ than 50% sure it _is_ the right fix.

**Comment by bogdan-iancu** [comment]
OK :) - may I ask what are your standing concerns ?

**Comment by wdoekes** [comment]
.. and a colleague reports that he is now (with your patch) unable to reproduce the dupes, while he _was able to_ before using a stock opensips (with the full 16 bits crc, not 4 bits like I used).

I think we can call that success :+1: 

(I haven't tested it myself, but the code changes looked good/sane.)

So, you fixed the **To-tag collision**, and the **Don't stop processing on failure**. That leaves the **Constraint [mismatch with the DB schema]**. Do you want that in a separate ticket so you can de-prioritize it?

**Comment by bogdan-iancu** [comment]
@wdoekes , great news ! Yes, let's fork a separate ticket for the remaining part and have this one closed. I will also backport this fix to 2.4 also ;)

**Comment by wdoekes** [comment]
I added #1772.

Thanks!

(For the record: 4e29ce92 (3.0), 7e24e5a9 (2.4) fix problem 3. And a9d8d437 (2.4) fixes problem 1.)
